File: WEEK6/Js/bj3_11438.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

let lineIdx = 0;
const N = Number(input[lineIdx++]);

let K = 0;
let temp = 1;
while (temp <= N) {
    temp *= 2;
    K++;
}

const adj = Array.from({ length: N + 1 }, () => []);
const parent = Array.from({ length: K + 1 }, () => new Int32Array(N + 1));
const depth = new Int32Array(N + 1);
const visited = new Uint8Array(N + 1);

for (let i = 0; i < N - 1; i++) {
    const [a, b] = input[lineIdx++].split(' ').map(Number);
    adj[a].push(b);
    adj[b].push(a);
}

const stack = [[1, 0]];
visited[1] = 1;

while (stack.length) {
    const [node, d] = stack.pop();
    depth[node] = d;

    for (const next of adj[node]) {
        if (!visited[next]) {
            visited[next] = 1;
            parent[0][next] = node;
            stack.push([next, d + 1]);
        }
    }
}

for (let i = 1; i <= K; i++) {
    for (let j = 1; j <= N; j++) {
        parent[i][j] = parent[i - 1][parent[i - 1][j]];
    }
}

const M = Number(input[lineIdx++]);
const output = [];

function lca(a, b) {
    if (depth[a] < depth[b]) [a, b] = [b, a];

    for (let i = K; i >= 0; i--) {
        if (Math.pow(2, i) <= depth[a] - depth[b]) {
            a = parent[i][a];
        }
    }

    if (a === b) return a;

    for (let i = K; i >= 0; i--) {
        if (parent[i][a] !== parent[i][b]) {
            a = parent[i][a];
            b = parent[i][b];
        }
    }

    return parent[0][a];
}

for (let i = 0; i < M; i++) {
    const [a, b] = input[lineIdx++].split(' ').map(Number);
    output.push(lca(a, b));
}

console.log(output.join('\n'));

File: WEEK6/python/bj3_11438.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline

n = int(input())
k_max = 0
temp = 1
while temp <= n:
    temp *= 2
    k_max += 1

adj = [[] for _ in range(n + 1)]
parent = [[0] * (n + 1) for _ in range(k_max + 1)]
depth = [0] * (n + 1)
visited = [False] * (n + 1)

for _ in range(n - 1):
    a, b = map(int, input().split())
    adj[a].append(b)
    adj[b].append(a)


def dfs(node, d):
    visited[node] = True
    depth[node] = d
    for next_node in adj[node]:
        if not visited[next_node]:
            parent[0][next_node] = node
            dfs(next_node, d + 1)


dfs(1, 0)

for i in range(1, k_max + 1):
    for j in range(1, n + 1):
        parent[i][j] = parent[i - 1][parent[i - 1][j]]


def lca(a, b):
    if depth[a] < depth[b]:
        a, b = b, a

    for i in range(k_max, -1, -1):
        if (1 << i) <= depth[a] - depth[b]:
            a = parent[i][a]

    if a == b:
        return a

    for i in range(k_max, -1, -1):
        if parent[i][a] != parent[i][b]:
            a = parent[i][a]
            b = parent[i][b]

    return parent[0][a]


m = int(input())
for _ in range(m):
    a, b = map(int, input().split())
    print(lca(a, b))
